### PR TITLE
Prevent ArrayBuilder capacity overflow/infinite looping in ArrayBuilder.ensureSize(size:Int).

### LINKED ISSUE #12617: ArrayBuilder should use smart sizing from ArrayBuffer
## References

Follow-up to https://github.com/scala/bug/issues/12464

## Problem

`ArrayBuilder` doesn't attempt smart resizing as the size grows large. It still just tries naive doubling of size.

When user offers a `sizeHint`, that is probably a clue to allocate a result buffer (and copy if necessary).

Otherwise, the implementer might choose to assemble partial buffers at `result` time, as suggested on chat.

### PR DESCRIPTION
**Pull request by dragonfly-ai**
Added:
  - capacity limit to ArrayBuilder
  - guards against exceeding capacity
  - guard against capacity overflow which triggers an infinite loop in ArrayBuilder.ensureSize
  - capacity limit guards to ArrayBuilder.ofUnit
  - Unit test to exercise and test the behavior of ArrayBuilder.ensureSize as it approaches the capacity limits.

Fixes https://github.com/scala/bug/issues/12617

### CODE REVIEW

**Comment by dragonfly-ai** [comment]
This PR completes or at least continues from [this one](https://github.com/scala/scala/pull/10442).

**Comment by dragonfly-ai** [comment]
Without this bug fix, trying to grow `ArrayBuilder` larger than 1073741824 silently enters an infinite loop.

This PR makes it possible for users to grow `ArrayBuilder` as large as `Int.MaxInt - 8`.

File: src/library/scala/collection/mutable/ArrayBuilder.scala [hunk 1]
```diff
@@ -31,16 +31,29 @@ sealed abstract class ArrayBuilder[T]
...
   override def knownSize: Int = size
 
+  private final val MaxArraySize:Int = 2147483639
+
   protected[this] final def ensureSize(size: Int): Unit = {
     if (capacity < size || capacity == 0) {
-      var newsize = if (capacity == 0) 16 else capacity * 2
-      while (newsize < size) newsize *= 2
-      resize(newsize)
+      if (size <= 1073741824) {
```
Comment (lrytz) [inline src/library/scala/collection/mutable/ArrayBuilder.scala]: If `capacity` is `1073741823` and size is `1073741824`, then we get `newSize = 2147483646`, which is larger than `MaxArraySize`

File: src/library/scala/collection/mutable/ArrayBuilder.scala [hunk 2]
```diff
@@ -31,16 +31,29 @@ sealed abstract class ArrayBuilder[T]
 
   override def knownSize: Int = size
 
+  private final val MaxArraySize:Int = 2147483639
```
Comment (lrytz) [inline src/library/scala/collection/mutable/ArrayBuilder.scala]: We should use `PStatics.VM_MaxArraySize` instead of introducing another constant. I would agree changing it to `Int.MaxValue - 8`.

Previous discussion: https://github.com/scala/scala/pull/9819#discussion_r762543984

File: src/library/scala/collection/mutable/ArrayBuilder.scala [hunk 3]
```diff
@@ -31,16 +31,29 @@ sealed abstract class ArrayBuilder[T]
...
+        ) else {
+          var newSize = if (capacity == 0) 16 else capacity * 2
+          while (newSize < size) newSize *= 2
+          resize(newSize)
+        }
+      } else if (capacity < MaxArraySize && size <= MaxArraySize) {
+        resize(MaxArraySize)
+      } else throw new Exception(
+        s"ArrayBuilder cannot store more than $MaxArraySize elements and cannot ensure a size of $size."
+      )
```
Comment (lrytz) [inline src/library/scala/collection/mutable/ArrayBuilder.scala]: I would prefer if we could align the code here with https://github.com/scala/scala/blob/v2.13.11/src/library/scala/collection/mutable/ArrayBuffer.scala#L321-L325. It makes maintenance much easier in the long turn.

The implementation in `ArrayBuffer` doubles the current size, but if that's still too small it jumps to the exact requested size. Discussion about this is here: https://github.com/scala/bug/issues/12464#issuecomment-970841156.

File: src/library/scala/collection/mutable/ArrayBuilder.scala (hunk 2 above)
Comment (dragonfly-ai) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: @Ichoran suggested that `PStatics` might not be available on some JVMs, but I definitely agree about using `PStatics.VM_MaxArraySize` and changing it to `Int.MaxValue - 8`.

Parallel discussion here: [#10442 (comment)](https://github.com/scala/scala/pull/10442#discussion_r1239303758)

@sjrd suggested using the private final val instead, in [this comment](https://github.com/scala/scala/pull/10442#discussion_r1241034211).

File: src/library/scala/collection/mutable/ArrayBuilder.scala (hunk 1 above)
Comment (dragonfly-ai) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: Oh, you're right.  I was thinking of capacity as a power of two, but users can disrupt that by setting `sizeHint(size:Int)`.

File: src/library/scala/collection/mutable/ArrayBuilder.scala (hunk 3 above)
Comment (dragonfly-ai) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: I can work on that.  You've linked to private methods; do you have a suggestion for where the shared logic should live?

Comment (lrytz) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: I don't have strong opinions, you could simply change `resizeUp` to be `private[mutable]` and leave it where it is in `object ArrayBuffer`.

File: src/library/scala/collection/mutable/ArrayBuilder.scala (hunk 2 above)
Comment (lrytz) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: Maybe there was some misunderstanding? `PStatics` is part of the Scala standard library so it should be available everywhere.

File: src/library/scala/collection/mutable/ArrayBuilder.scala (hunk 3 above)
Comment (som-snytt) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: I did not intend to ignore this PR, I was waiting for dust to settle. I'll contribute eyeballs after I unbreak unused imports on my own PR.

Comment (dragonfly-ai) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: > I don't have strong opinions, you could simply change `resizeUp` to be `private[mutable]` and leave it where it is in `object ArrayBuffer`.

I'll try that.

Comment (dragonfly-ai) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: Are we sure that won't cause any binary incompatibility?

Comment (som-snytt) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: Binary compat is about public API, what users might link, exclusive of inlining.

Comment (dragonfly-ai) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: @lrytz, I have concerns about the way `ArrayBuffer` handles resizing.

```scala
if (targetLen > Int.MaxValue) throw new Exception(s"Collections cannot have more than ${Int.MaxValue} elements")
IterableOnce.checkArraySizeWithinVMLimit(targetLen.toInt) // safe because `targetSize <= Int.MaxValue`
```
Why not just replace these two lines with the relevant internals of `checkArraySizeWithinVMLimit`:
```scala
if (targetLen > scala.runtime.PStatics.VM_MaxArraySize) throw new Exception(
  s"Size of array-backed collection exceeds VM array size limit of ${scala.runtime.PStatics.VM_MaxArraySize}"
)
```

The current way seems overly complicated and implies that the user can allocate an `Array`, `ArrayBuffer`, or `ArrayBuilder` of size `Int.MaxValue` when the true limit, `scala.runtime.PStatics.VM_MaxArraySize`, is smaller.

More importantly, anything involving `Long` will cause significant performance problems in Scala.js.  Since every single call to `addAll` or `addOne` will perform several operations on Long, this approach might be a mistake.  Maybe we ought to detect overflow without casting sizes to long values?

Scala.js has [its own version of ArrayBuilder](https://github.com/scala-js/scala-js/blob/main/scalalib/overrides-2.12/scala/collection/mutable/ArrayBuilder.scala), but wouldn't changes to this version necessarily have to cascade to there?

Comment (lrytz) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: Simplifying and unifying sounds good to me
  - Use `addExact` instead of converting to `Long`?
  - Only check against `VM_MaxArraySize`; you could keep `checkArraySizeWithinVMLimit` for that, it's marked `@inline` and we build Scala releases with the optimizer enabled. But either way is fine with me.

For Scala.js, ping @sjrd

Comment (sjrd) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: In Scala.js, `Long`s are indeed expensive when they cross method boundaries. If you use `Long`s locally within a method they're OK.

Regardless, if you're using `Long`s only to test that the result of `a + b` does not overflow and you know that `a >= 0 && b >= 0`, the most efficient remains to test whether the result is `< 0`.

Comment (dragonfly-ai) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: @lrytz I'd like to follow sjrd on this and rewrite the long values out of the resizing logic.

Would you oppose?

Comment (lrytz) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: No, sounds good to me 👍

**Comment by lrytz** [comment]
Assigned to 2.13.13, but this can still go into 2.13.12 if it's ready. I just wanted to have the 2.13.12 milestone as a to do / to review list, and this PR is not a blocker.

File: src/library/scala/collection/mutable/ArrayBuilder.scala (hunk 2 above)
Comment (Ichoran) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: @lrytz -
```scala
Welcome to Scala 3.3.0 (17.0.7, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.
                                                                 
scala> scala.runtime.PStatics.VM_MaxArraySize
-- [E173] Reference Error: -----------------------------------------------------
1 |scala.runtime.PStatics.VM_MaxArraySize
  |^^^^^^^^^^^^^^^^^^^^^^
  |object PStatics cannot be accessed as a member of runtime.type from module class rs$line$2$.
1 error found
```

File: src/library/scala/collection/mutable/ArrayBuilder.scala (hunk 3 above)
Comment (Ichoran) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: When using doubling to come up with a new size, you can be careful with your operations to avoid overflow.  For instance, instead of testing `2*n >= maxSize` you can test `n >= maxSize/2`.  Or instead of `n + m >= maxSize` you can use `m >= maxSize - n` (where `n` is known to be between `0` and `maxSize`).

Once you do the check that it's in range, you still have to calculate the correct value (`2*n` or `n+m`), but integer math is very inexpensive, so that's no big deal.  For a core library routine, spending the effort to do it this way (correctly) is probably worth it.

(Another option is to do the `2*n` or `n + m` but then the test is `x >= maxSize || x < 0` to catch the overflow case, assuming that your input values are nonnegative.)

File: src/library/scala/collection/mutable/ArrayBuilder.scala (hunk 2 above)
Comment (lrytz) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: It's `private[scala]`, so not public API, but available for use within the standard library.

Comment (Ichoran) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: I'm skeptical that this makes sense; anything that is broad enough for use by the entire Scala library but which isn't a compiler-specific detail is something that's likely to be needed in user-space too.  But I guess that can be fixed later.

Comment (dragonfly-ai) [reply src/library/scala/collection/mutable/ArrayBuilder.scala]: This ought to be public, and ought to live somewhere more visible.

It really ought to be in Java, though.

**Comment by dragonfly-ai** [comment]
nothing new committed yet.  just refreshed.

**Comment by som-snytt** [comment]
FYI it seems 2.13.12 is end of summer, so there is plenty of time and no rush. However, since you began this effort in summer 2022, you will need to honor the tradition of taking the reviewers out to an anniversary dinner. I don't know the rule when the collections collective is listed as a reviewer. Maybe pick a nice buffet place?

Happy Anniversary! https://github.com/scala/bug/issues/12617#issuecomment-1180709064 I almost missed it.

**Comment by SethTisue** [comment]
@dragonfly-ai interested in returning to this?

**Comment by dragonfly-ai** [comment]
> @dragonfly-ai interested in returning to this?

It's quite the design challenge.

Consolidate code that's very different into a smaller set of common methods without changing the public interfaces at all and also in a way in which the resizing logic can be tested without actually resizing anything.

I haven't yet found a way to address all of these concerns with a single design.

Do you have anything in mind, or any suggestions?

**Comment by SethTisue** [comment]
I'll re-mention @scala/collections, the last mention was before we added a bunch of new volunteers to the team.

**Comment by lrytz** [comment]
> > @dragonfly-ai interested in returning to this?
> 
> It's quite the design challenge.

I think we were on a good track in the previous discussion: https://github.com/scala/scala/pull/10448/files#r1246274512

**Comment by dragonfly-ai** [comment]
Just syncing the fork.  Nothing new to review.

**Comment by dragonfly-ai** [comment]
OK, this is a sincere commit!  Thank you all for your patience!

**Comment by lrytz** [comment]
@dragonfly-ai I rebased on top of current 2.13.x (there was a conflict). I squashed your commits into one and added my cleanups from reviewing. Please take a look.

Note that this PR is based off your `2.13.x` branch. The usual workflow is to create a new branch for each PR and keep 2.13.x in sync with upstream.

**Comment by dragonfly-ai** [comment]
Thanks for squashing and rebasing.  I'm a source control neophyte.

Good call to move checkArraySizeWithinVMLimit out of IterableOnce.  I wanted to do that, too since nothing else uses it, but felt shy about moving around other people's contributions.

Really, though,  Maybe PStatics shouldn't exist either.  It only holds one value which only gets used by ArrayBuilder and ArrayBuffer and their tests.

I also like what you did to the comments and cleanup.  It helped me get a sense of scala codebase style norms.  I tried to respect them, but sorry if I brought too much of my own style.

File: src/library/scala/collection/mutable/ArrayBuffer.scala [hunk 4]
```diff
@@ -314,24 +306,35 @@ object ArrayBuffer extends StrictOptimizedSeqFactory[ArrayBuffer] {
...
   /**
+   * The increased size for an array-backed collection.
+   *
    * @param arrayLen  the length of the backing array
    * @param targetLen the minimum length to resize up to
-   * @return -1 if no resizing is needed, or the size for the new array otherwise
+   * @return
+   *   - `-1` if no resizing is needed, else
+   *   - `VM_MaxArraySize` if `arrayLen` is too large to be doubled, else
+   *   - `max(targetLen, arrayLen * 2, , DefaultInitialSize)`.
```
Comment (som-snytt) [inline src/library/scala/collection/mutable/ArrayBuffer.scala:323]: next tweak could delete a comma

File: test/junit/scala/collection/mutable/ArrayBufferTest.scala [hunk 5]
```diff
@@ -387,28 +388,31 @@ class ArrayBufferTest {
...
     assertEquals(ArrayBuffer.DefaultInitialSize, resizeUp(7, 10))
-    assertEquals(Int.MaxValue - 2, resizeUp(Int.MaxValue / 2, Int.MaxValue / 2 + 1))  // was: ok
-    assertEquals(Int.MaxValue - 2, resizeUp(Int.MaxValue / 2, Int.MaxValue / 2 + 2))  // was: ok
-    assertEquals(-1, resizeUp(Int.MaxValue / 2 + 1, Int.MaxValue / 2 + 1))  // was: wrong
-    assertEquals(Int.MaxValue - 2, resizeUp(Int.MaxValue / 2 + 1, Int.MaxValue / 2 + 2))  // was: hang
-    assertEquals(Int.MaxValue - 2, resizeUp(Int.MaxValue / 2, Int.MaxValue - 2))
+    assertEquals(VM_MaxArraySize, resizeUp(Int.MaxValue / 2, Int.MaxValue / 2 + 1)) // `MaxValue / 2` cannot be doubled
+    assertEquals(VM_MaxArraySize / 2 * 2, resizeUp(VM_MaxArraySize / 2, VM_MaxArraySize / 2 + 1)) // `VM_MaxArraySize / 2` can be doubled
+    assertEquals(VM_MaxArraySize, resizeUp(Int.MaxValue / 2, Int.MaxValue / 2 + 2))
+    assertEquals(-1, resizeUp(Int.MaxValue / 2 + 1, Int.MaxValue / 2 + 1)) // no resize needed
```
Comment (som-snytt) [inline test/junit/scala/collection/mutable/ArrayBufferTest.scala:399]: these couple of asserts are where I need the comment that says, "This regression test is for historical purposes, like all regression tests. It used to { hang, give a wrong result, throw, whatever }." That info is useless in so far as it is merely historical and not of theoretical interest. Put another way, I might say now, "What a stupid test, it should be a scalacheck, I can't even see what sort of boundary condition it thinks it is probing."

Leaving this comment as meta-discussion. It's not actionable.

**Comment by som-snytt** [comment]
@dragonfly-ai there is a comment on PStatics that it is temporary. I guess the P is for private? I thought it was something mysterious. Also, the final val is an inline constant, so it merely has to reside in an interface somewhere.

**Comment by dragonfly-ai** [comment]
@som-snytt Maybe after SIP-51, we can remove dumb tests and PStatics without fear of angering the bincompat?

File: src/library/scala/collection/mutable/ArrayBuilder.scala [hunk 6]
```diff
@@ -34,15 +35,11 @@ sealed abstract class ArrayBuilder[T]
   override def knownSize: Int = size
 
   protected[this] final def ensureSize(size: Int): Unit = {
```
Comment (som-snytt) [inline src/library/scala/collection/mutable/ArrayBuilder.scala:37]: it should be called `ensureCapacity`. I've already had coffee, but it made me blink twice. Maybe you think it's too obvious to make the distinction between size and capacity obviouser.

Comment (dragonfly-ai) [reply src/library/scala/collection/mutable/ArrayBuilder.scala:37]: I agree with you, so much so that: in NArr, I'm implementing `NArrayBuilder` and definitaly named it `ensureCapacity`.

Isn't it too late to rename it now, though, after 20 years?

File: test/junit/scala/collection/mutable/ArrayBuilderTest.scala [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+      def next(): Unit = ()
+    })
+
+    // reached maximum size without entering an infinite loop?
+    assertEquals(ab.length, VM_MaxArraySize)
+
+    // expect an exception when trying to grow larger than maximum size by addOne
+    assertThrows[Exception](ab.addOne(()), _.endsWith("Requested length: 2147483640; current length: 2147483639"))
+
+    val arr = Array[Unit]((), (), (), (), (), (), (), (), (), (), (), ())
```
Comment (som-snytt) [inline test/junit/scala/collection/mutable/ArrayBuilderTest.scala:27]: `Array.fill(12)` to remind myself. Also `Array.fill(12)()` is correctly linted. L.A. Phil is the philharmonic. MPhil is the degree.

**Review by som-snytt — APPROVED** [review]
Another nice lrytz review experience. This is just supplementary meta-appreciation. Addition and multiplication by 2 are the limit of my math. Nice job!

File: src/library/scala/collection/mutable/ArrayBuffer.scala [hunk 8]
```diff
@@ -307,24 +306,29 @@ object ArrayBuffer extends StrictOptimizedSeqFactory[ArrayBuffer] {
...
    */
   private[mutable] def resizeUp(arrayLen: Int, targetLen: Int): Int = {
-    // Hybrid
     if (targetLen > 0 && targetLen <= arrayLen) -1
     else {
-      IterableOnce.checkArraySizeWithinVMLimit(targetLen) // safe because `targetSize <= Int.MaxValue`
-
-      math.min(
-        if (targetLen > (Int.MaxValue >> 1)) VM_MaxArraySize
-        else math.max(targetLen, math.max(arrayLen << 1, DefaultInitialSize)),
```
Comment (dragonfly-ai) [inline src/library/scala/collection/mutable/ArrayBuffer.scala]: ok, / and * instead of >> 1 and << 1.  Now I know.

**Comment by som-snytt** [comment]
> we can remove dumb tests

I don't think you're ever allowed to remove stupid tests, only wrong ones. That is for the reason I cited, that without a comment, I don't even know what it was testing, and maybe I'm the stupid one, but we know it was testing something that was once unforeseen. Or let me retract that, because occasionally one will notice a test that actually tests nothing: it may have progressed from testing to non-testing, as opposed to progressing from failing to succeeding.

File: src/library/scala/collection/mutable/ArrayBuffer.scala (hunk 8 above)
Comment (som-snytt) [reply src/library/scala/collection/mutable/ArrayBuffer.scala]: That's a standard thing. Shift has an aura of metal, but compilers know about powers of 2. Even I know how to double a number, at least up to ten fingers. So we prefer the notation that is easier to read.

Comment (Ichoran) [reply src/library/scala/collection/mutable/ArrayBuffer.scala]: An exception is when you need to do correct division despite overflow.  `(a + b)/2` fails with signed math if `a` and `b` are sufficiently large (positive), but `(a + b) >>> 1` works correctly because signed and unsigned `a + b` have the same bit pattern, and `>>> 1` is a shift-as-if-unsigned.  Which necessarily brings it back into signed range.

This is WAY easier than the other logic you often need to do to avoid this edge case.

**Review by Ichoran — APPROVED** [review]
LGTM!  The core logic all passes my visual inspection.

**Comment by som-snytt** [comment]
```
git fetch upstream
git checkout upstream/2.13.x -b issue/1234
git push origin issue/1234
```
in case that wasn't clear.

**Comment by som-snytt** [comment]
Merging with Ichoran :eyes: because lrytz is too close and has been working on it for too long and it is the holiday season.

File: test/junit/scala/collection/mutable/ArrayBuilderTest.scala [hunk 9]
```diff
@@ -1,40 +1,35 @@
...
-    while (i < MaxArraySize) {
-      ab.addOne(())
-      i += 1
-    }
+    // ArrayBuilder.ofUnit.addAll doesn't iterate if the iterator has a `knownSize`
+    ab.addAll(new Iterator[Unit] {
+      override def knownSize: Int = VM_MaxArraySize
+      def hasNext: Boolean = true
+      def next(): Unit = ()
+    })
```
Comment (lrytz) [inline test/junit/scala/collection/mutable/ArrayBuilderTest.scala:19]: For those still watching, here's an interesting bit.

The while loop variant was running fine, actually very quickly because the JVM optimized the loop away. But as soon as I attached a debugger with a breakpoint somewhere below `addOne`, the VM had to de-optimize and the test didn't finish (or I guess it would have, eventually).

Comment (dragonfly-ai) [reply test/junit/scala/collection/mutable/ArrayBuilderTest.scala:19]: Is this a concern, or can we just not set breakpoints below `addOne`?

It seemed strange that no `ArrayBuilderTest` existed prior to this merge until I tried to write them and found that only `ofUnit` can have stress tests because it doesn't allocate any memory.  Every other builder overwhelms the heap and takes forever to run.

The `ArrayBuilderTest` we have now looks pretty bare except in comparison to the nothing that it replaced.

Should we trust the 20 year old implementations as battle hardened if not tested, or add more tests?

Comment (som-snytt) [reply test/junit/scala/collection/mutable/ArrayBuilderTest.scala:19]: The old joke is that the patient says doctor, it hurts when I do this. The doctor says then don't do that.

Collections still has a test problem, but arguably array ops would not work if array builder broke. Fortunately, array builder is simple (albeit the source is cut/paste for the special subclasses). For the purpose of building an array, it is preferable to array buffer, which is much more complex. Usually, you know how big the array must be, and growing it is an edge case. The convenience is in initializing the array.

Comment (lrytz) [reply test/junit/scala/collection/mutable/ArrayBuilderTest.scala:19]: Tests are always welcome, there are certainly under-tested collections in the repo.

But yeah, we cannot just have tests that allocate Int.MaxValue elements, so testing for edge cases there is not easy.

> Is this a concern, or can we just not set breakpoints below `addOne`?

Note that I rewrote the test to no longer use a while loop.

**Comment by SethTisue** [comment]
In the community build, this caused test failures in [scala-collection-laws](https://github.com/scala/scala-collection-laws)

the full log of all the failures is at https://scala-ci.typesafe.com/job/scala-2.13.x-jdk17-integrate-community-build/1554/artifact/logs/scala-collection-laws-build.log

but the simplest reproducer I know of is very simple:

```
scala> collection.mutable.ArrayBuffer[Int]().dropRight(1)
java.lang.Exception: Overflow while resizing array of array-backed collection. Requested length: -1; current length: 16; increase: -17
  at scala.collection.mutable.ArrayBuffer$.resizeUp(ArrayBuffer.scala:313)
  at scala.collection.mutable.ArrayBuffer$.scala$collection$mutable$ArrayBuffer$$ensureSize(ArrayBuffer.scala:338)
  at scala.collection.mutable.ArrayBuffer.ensureSize(ArrayBuffer.scala:69)
  at scala.collection.mutable.ArrayBuffer$$anon$1.sizeHint(ArrayBuffer.scala:304)
```

the full set of failures can be reproduced by cloning scala-collection-laws, starting sbt, doing `set Global / resolvers += "nightlies" at "https://scala-ci.typesafe.com/artifactory/scala-integration/"` then `++2.13.13-bin-97c2515!` then `Test/compile` then `laws/run`

@dragonfly-ai can I interest you in investigating...?

I do understand it's almost Christmas. please let us know if you think you could look soon; or if you think you could look in early January; or if you don't expect to have time to look at all, in which case no worries, and one of us will take it over

**Comment by som-snytt** [comment]
:notes: Don't they know it's Christmas? :notes:

**Comment by som-snytt** [comment]
PR per the link.

**Comment by dragonfly-ai** [comment]
> you could look soon

Sorry, just saw this.

Did som-snytt already fix it?

**Comment by SethTisue** [comment]
@dragonfly-ai It seems so, but perhaps you could review his changes at https://github.com/scala/scala/pull/10642

**Comment by dragonfly-ai** [comment]
The diffs seem reasonable upon first reading, but I fear I'd need yet another year to complete any useful deeper analysis.

**Comment by som-snytt** [comment]
@dragonfly-ai I didn't intend to steal your educational experience. It was surprisingly educational. But it also was not something subtle about your fix, thank goodness. And it was a nice change of pace from another PR that was getting me down.

@SethTisue thanks for laws instructions. I followed the project page. I'd forgotten I was running the tests in September for another PR about slice something something.

**Comment by SethTisue** [comment]
regression: https://github.com/scala/bug/issues/12971

**Comment by dragonfly-ai** [comment]
I feel responsible and can start digging into this on Monday.

If anybody else can get it sooner, no sense waiting up for me.

**Comment by He-Pin** [comment]
@dragonfly-ai https://github.com/scala/scala/pull/10722/
